### PR TITLE
9.0: FEATURE: Put initiating user id and creation date into event metadata

### PR DESCRIPTION
**Pull request by bwaidelich**
Resolves: #3874

### CODE REVIEW

File: Neos.ContentRepository.Core/Classes/ContentRepository.php [hunk 1]
```diff
@@ -75,6 +82,19 @@ public function handle(CommandInterface $command): CommandResult
...
+        $eventsToPublish = new EventsToPublish(
+            $eventsToPublish->streamName,
+            Events::fromArray($eventsToPublish->events->map(function (EventInterface|DecoratedEvent $event) {
+                $metadata = $event->eventMetadata->value;
+                $metadata['initiatingUserId'] = $this->initiatingUserId->value;
+                $metadata['initiatingTimestamp'] = (new \DateTimeImmutable())->format(\DateTimeInterface::ATOM);
+                return DecoratedEvent::withMetadata($event, EventMetadata::fromArray($metadata));
+            })),
+            $eventsToPublish->expectedVersion,
+        );
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/ContentRepository.php:96]: Just a PoC, the event decoration mechanism is still a bit quirky

File: Neos.ContentRepository.Core/Classes/Feature/ContentStreamCreation/Event/ContentStreamWasCreated.php [hunk 2]
```diff
@@ -28,23 +27,18 @@ final class ContentStreamWasCreated implements EventInterface
...
         );
     }
 
     public function jsonSerialize(): array
     {
-        return [
-            'contentStreamId' => $this->contentStreamId,
-            'initiatingUserId' => $this->initiatingUserId,
-        ];
+        return get_object_vars($this);
```
Comment (skurfuerst) [inline Neos.ContentRepository.Core/Classes/Feature/ContentStreamCreation/Event/ContentStreamWasCreated.php:42]: hm, can we keep this explicit? :) I feel this is too dangerous to have it implicit (same for the other changes)

File: Neos.ContentRepository.Core/Classes/ContentRepository.php (hunk 1 above)
Comment (bwaidelich) [reply Neos.ContentRepository.Core/Classes/ContentRepository.php:96]: I'll be working on that part today

File: Neos.ContentRepository.Core/Classes/Feature/ContentStreamCreation/Event/ContentStreamWasCreated.php (hunk 2 above)
Comment (bwaidelich) [reply Neos.ContentRepository.Core/Classes/Feature/ContentStreamCreation/Event/ContentStreamWasCreated.php:42]: As discussed, `get_object_vars()` is not really less explicit, it just makes all props of the event part of their payload which makes sense for events IMO and I think the risk of a typo is greater with the repetition

**Comment by bwaidelich** [comment]
FYI: I plan to adjust the remaining code and tests later today

File: Neos.ContentRepository.Core/Classes/ContentRepository.php [hunk 3]
```diff
@@ -75,9 +82,37 @@ public function handle(CommandInterface $command): CommandResult
...
         $eventsToPublish = $this->commandBus->handle($command, $this);
 
+        // add "initiatingUserId" and "initiatingTimestamp" metadata to all events
+        // TODO: cleanup
+        $eventsToPublish = new EventsToPublish(
+            $eventsToPublish->streamName,
+            Events::fromArray($eventsToPublish->events->map(function (EventInterface|DecoratedEvent $event) {
+                $metadata = $event instanceof DecoratedEvent ? $event->eventMetadata->value : [];
+                $metadata['initiatingUserId'] = $this->initiatingUserId->value;
+                $metadata['initiatingTimestamp'] = (new \DateTimeImmutable())->format(\DateTimeInterface::ATOM);
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/ContentRepository.php:92]: Consider already existing metadata and don't override (= copy user id & timestamp from existing events)

Comment (bwaidelich) [reply Neos.ContentRepository.Core/Classes/ContentRepository.php:92]: TODO: Use single DateTime instance

Comment (bwaidelich) [reply Neos.ContentRepository.Core/Classes/ContentRepository.php:92]: TODO: Add doc comment

**Pull request merged**